Move functor-level GEMM routines to KokkosBlas

mzuzek opened the pull request:
@lucbv @fnrizzi

This PR moves GEMM 

Status:
* Implementation:
  - [x] move implementations of `SerialGemm`, `TeamGemm` and `TeamVectorGemm` from `KokkosBatched` to `KokkosBlas` (`BatchedSerialGemm` stays in batched);
  - [x] move MKL implementation to dedicated header;
* Interface:
  - [x] move selective `Gemm` interface (dispatching between `SerialGemm` and `TeamGemm`) from `KokkosBatched` to `KokkosBlas`;
* Tests:
  - [x] move unit tests for functor-level routines;

Commit:
GEMM: Move Serial, Team and TeamVector implementations to KokkosBlas

Commit:
GEMM: move unit tests to Blas

fnrizzi commented:
just my opnion, but i think it would make sense to keep this PR only about moving things around, and then address the implementation of ConjTranspose and scalar stuff in a separate PR

Commit:
GEMM: connect TeamVectorGemm to the selective interface

Commit:
GEMM: move MKL implementation of SerialGemm to dedicated TPL header

Commit:
GEMM: implicit MemberType

mzuzek commented on src/blas/KokkosBlas3_gemm_inner_fix.hpp:
```diff
@@ -0,0 +1,166 @@
+
+#include "KokkosBlas3_serial_gemm_inner_fixa_impl.hpp"
+// TODO: fix compilation errors InnerGemmFixB (not used internally, not tested)
+// #include "KokkosBlas3_serial_gemm_inner_fixb_impl.hpp"
```
Compiler errors in this header were not detected before, because it was not used anywhere (no internal need for `InnerGemmFixB` and no tests for it).

mzuzek commented on src/blas/KokkosBlas3_gemm.hpp:
```diff
@@ -256,6 +256,105 @@ void gemm(const char transA[], const char transB[],
+///
+template <typename ArgTransA, typename ArgTransB, typename ArgMode,
+          typename ArgAlgo = Algo::Gemm::Default>
+struct Gemm {
```
@lucbv Should this interface be placed in `KokkosBlas::Experimental` namespace ? (it wasn't in `KokkosBatched`)

mzuzek commented:
@lucbv @fnrizzi 
I reviewed the headers and interfaces trying to apply what we've discussed, so let me open the PR for reviews.

For the following tasks, I'll create follow-up PRs:
* implement `ConjTranspose` (follow-up: cover in tests);
* implement mixed input scalar types (value types in `A`, `B` and `C` matrices);
* extend unit tests to cover `Algo::Gemm::CompactMKL` (MKL implementation that works on SIMD vectors);

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

lucbv reviewed (commented):
You are defining a macro for COMPACT_MKL multiple times but it should really be centralized instead.

Commit:
GEMM: bring back batched interfaces for backward compatibility

Commit:
MKL: move utils to common header + fix macro duplication

mzuzek commented on src/batched/dense/KokkosBatched_Gemm_Decl.hpp:
```diff
@@ -51,73 +51,6 @@
-///
-
-template <typename ArgTransA, typename ArgTransB, typename ArgAlgo>
-struct SerialGemm {
```
Good point, thank you. Fixed on https://github.com/kokkos/kokkos-kernels/pull/1519/commits/11f135dd1762b01f729a6aa54eb5bc9dfa2ec75d.
> I've tested these restored interfaces with batched GEMM tests from `develop`.

mzuzek commented on src/batched/dense/KokkosBatched_Gemm_Decl.hpp:
```diff
@@ -51,73 +51,6 @@
-
-template <typename MemberType, typename ArgTransA, typename ArgTransB,
-          typename ArgAlgo>
-struct TeamGemm {
```
fixed with https://github.com/kokkos/kokkos-kernels/commit/11f135dd1762b01f729a6aa54eb5bc9dfa2ec75d

mzuzek commented on src/batched/dense/KokkosBatched_Gemm_Decl.hpp:
```diff
@@ -51,73 +51,6 @@
-
-template <typename MemberType, typename ArgTransA, typename ArgTransB,
-          typename ArgAlgo>
-struct TeamVectorGemm {
```
fixed with https://github.com/kokkos/kokkos-kernels/commit/11f135dd1762b01f729a6aa54eb5bc9dfa2ec75d

mzuzek commented:
@lucbv 
> You are defining a macro for COMPACT_MKL multiple times but it should really be centralized instead.

Thank you for pointing this out. On https://github.com/kokkos/kokkos-kernels/pull/1519/commits/34c50ea585c1df74031960c7dbfcacfcdc531c4e I've created `common/impl/KokkosKernels_MKLUtils.hpp` for common MKL utils used in blas and sparse (also fixed `common/src/impl` header path in CMake) and moved the duplicated macro there (also `#inlcude <mkl.h>` and `KOKKOSKERNELS_MKL_SAFE_CALL`). Let me know in case you'd wish a different location, name or content.

Commit:
GEMM: fix "batched" in names

mzuzek commented:
Added one trivial fix on 53c58bc55a325ed20c22c69b56a6fab541ad7ebf and rebased on `develop`.

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

kokkos-devops-admin commented:
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

lucbv commented:
Won't do at this time so closing this
